File: HW4/Q6/q6.m
import brml.*;
load("EMprinter.mat");

fuse = array(1, condp(rand(1, 2)));
drum = array(2, condp(rand(1, 2)));
toner = array(3, condp(rand(1, 2))); 
paper = array(4, condp(rand(1, 2)));
roller = array(5, condp(rand(1, 2))); 
burning = array([6 1], condp(rand(2, 2)));
quality = array([7 2 3 4], condp(rand(2, 2, 2, 2)));
wrinkled = array([8 1 4], condp(rand(2, 2, 2)));
multpages = array([9 4 5], condp(rand(2, 2, 2))); 
paperjam = array([10 1 5], condp(rand(2, 2, 2)));

% use EM algorithm to learn all CPTs of the network
pot = {fuse, drum, toner, paper, roller, burning, quality, wrinkled, multpages, paperjam};
pars.maxiterations = 10000;
[newpot, loglik] = EMbeliefnet(pot, x, pars);

% inference part of Q6
joint = condpot(multpots(newpot));
summedOut = sumpot(joint, [1 3 4 5 9 10]);
drumUnitProbs = condp(squeeze(summedOut.table(:, 1, 2, 1)));
fprintf("Prob. of drum unit problem: %.4f\n", drumUnitProbs(2));

File: HW4/Q6/q6.py
import numpy as np
import pandas as pd

from scipy.io import loadmat

class CPT():
    def __init__(self, name, parents):
        self.name = name
        self.parents = parents
        self.table = np.random.rand(2 ** len(parents))

    def get_MLE(self, sample):
        return

data = loadmat("EMprinter.mat")["x"]
variables = ["fuse", "drum", "toner", "paper", "roller"] + ["burning", "quality", "wrinkled", "mult.pages", "paperjam"] # diagnoses + faults

N, D = data.shape
EPSILON = 0.01
NUM_VARIABLES = 10

oldTables = [CPT("fuse", []), CPT("drum", []), CPT("toner", []), CPT("paper", []), CPT("roller", [])] + [CPT("burning", [0]), [CPT("quality", [1, 2, 3]), [CPT("wrinkled", [0, 3]), [CPT("mult.pages", [3, 4]), [CPT("paperjam", [0, 4])]
while diff < EPSILON: # check for convergence
    newTables = [None] * NUM_VARIABLES # initialize tables for next iteration



    # calculate diff to check for convergence
    diff = 0.0
    for i in range(NUM_VARIABLES):
        if i < 5:
            diff += abs(newTables[i] - oldTables[i])
        else:
            diff += np.linalg.norm(newTables[i] - oldTables[i], ord = 1)
